File: 2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points/2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    vector<int> nodesBetweenCriticalPoints(ListNode* head) {
        int Min = 100000, i = 1;
        int c[2] = {0, 0};
        auto prev = head, curr = head->next, nxt = head->next->next;
        auto isCrit = [&]() {
            auto x = prev->val, y = curr->val, z = nxt->val;
            return (x < y && y > z) || (x > y && y < z);
        };
        while (nxt) {
            if (isCrit()) {
                if (c[0]) Min = min(Min, i - c[c[1] > 0]);
                c[c[0] > 0] = i;
            }
            prev = curr; curr = nxt;
            nxt = nxt->next; i++;
        }
        if (c[1]) return {Min, c[1] - c[0]};
        return {-1, -1};
    }
};

File: 2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points/2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def nodesBetweenCriticalPoints(self, head: Optional[ListNode]) -> List[int]:
        def is_crit(x, y, z):
            return (y.val - x.val) * (y.val - z.val) > 0

        c = [0, 0]
        Min, i = inf, 1

        prev, curr, nxt = head, head.next, head.next.next        

        while nxt:
            if is_crit(prev, curr, nxt):
                if c[0]: Min = min(Min, i - c[c[1] > 0])
                c[c[0] > 0] = i

            prev, curr, nxt = curr, nxt, nxt.next
            i += 1

        return [[Min, c[1] - c[0]], [-1, -1]][not c[1]]